### ISSUE TITLE
Definition list entry with unresolved conref breaks publishing

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2701]
Let's say that in a DITA topic I have a definition list containing a list entry which conrefs to topics which do not resolve:

```xml
<dlentry conkeyref="lib-a/lib-1" conrefend="ignore.xml#lib-a/lib-3" id="dlentry_pvk_lww_fz">
  <dt id="dt_qvk_lww_fz" />
  <dd id="dd_rvk_lww_fz" />
</dlentry>
```
When publishing to PDF the generated XSL-FO is invalid so the publishing breaks with:

>[fop] [ERROR] net.sf.saxon.trans.XPathException: org.apache.fop.fo.ValidationException: "fo:block" is not a valid child of "fo:table-body"! (See position 8:112)
>[fop] [ERROR] Anttask - Error rendering fo file: C:\Users\radu_coravu\Desktop\dita-test1\dita-docs\ditamaps\temp\pdf\topic.fo <org.apache.fop.apps.FOPException: org.apache.fop.fo.ValidationException: "fo:block" is not a valid child of "fo:table-body"! (See position 8:112)
>[fop] net.sf.saxon.trans.XPathException: org.apache.fop.fo.ValidationException: "fo:block" is not a valid child of "fo:table-body"! (See position 8:112)>org.apache.fop.apps.FOPException: org.apache.fop.fo.ValidationException: "fo:block" is not a valid child of "fo:table-body"! (See position 8:112)
>[fop] net.sf.saxon.trans.XPathException: org.apache.fop.fo.ValidationException: "fo:block" is not a valid child of "fo:table-body"! (See position 8:112)
>[fop] 	at org.apache.fop.tools.anttasks.FOPTaskStarter.renderInputHandler(Fop.java:652)

The XSLT stylesheet `org.dita.pdf2\xsl\fo\commons.xsl` contains a template which outputs an error message block in the XSL-FO when a conref cannot be resolved:

```xml
<xsl:template match="*[@conref]" priority="99">
```
but as the `dl` element is rendered as an `fo:table` this template outputs an `fo:block` as a child of `fo:table-body`.

**Comment by robander** [comment]
I'm having trouble reproducing with latest `develop` code. It appears the conref that doesn't resolve is stripped out of the topic during `preprocess`, so that the `*[@conref]` rule in the FO code will never fire. I'm not sure if this is a change, or possibly you've got custom code in place that will preserve the element?

**Comment by robander** [comment]
Think I found a related issue with this. When I've got this invalid conref: `conref="#./missing"` I get a warning that the conref cannot resolve. When I've got a document that doesn't exist -- `conref="missing.dita#missing/invalid` -- I get a warning that the file can't be opened, but no message about the unresolved conref.

**Comment by raducoravu** [comment]
Somehow in this particular case the conreffed topic and element IDs were all there but they were in a folder structure above the folder where the main published DITA Map was located and the "generate.copy.outer" was set to default "1".
I think the XSLTs on the PDF or HTML side should no longer preoccupy or add errors in the generated document about unresolved conrefs, this is a job of the preprocessing stage which should issue warnings/errors in the console output. Especially because a missing conref is not a fatal error in DITA as there is always a fallback for it.

**Comment by robander** [comment]
The `preprocess` is broken for conref messages to a missing file right now - I've just about got a fix ready for it. 

I think if the broken conref markup flows through to PDF, highlighting the broken conref might be helpful - to help somebody locate it, or let a reviewer see it and realize the problem. I assume that was the idea behind the current processing, which is indeed broken for `@conref` in many locations. It shouldn't give a second message about the same thing though.

**Comment by robander** [comment]
The missing conref message (no / insufficient message when target doc unavailable) is resolved in 2.5 with https://github.com/dita-ot/dita-ot/commit/b466eea120a45c526963b22fdccd0d9681bd484a

Chatted with @jelovirt about this a bit and came around to thinking that the PDF code (or HTML5, or XHTML, etc) should not need special coding to handle conref -- the conref preprocess should handle that. PDF might still fail in this case, but a more general update to handle source errors (like source that has empty `<dlentry>`) would cover source errors + conref errors.

**Comment by robander** [comment]
More thoughts ...

- If an unresolved `@conref` on `<dlentry>` is removed ... we could end up with the same situation we'd have if all `<dlentry>` elements were filtered out of `<dl>`. For this, we could update `<dl>` processing to ensure it doesn't create `fo:table` without required children.
- If the unresolved `@conref` remains ... we circle around to the same issue. With unresolved `<dl conref="#./bad">`, we have the same issue - an element that gets to PDF code without required children.

So a potential fix here is to remove the rule for `@conref` and add more graceful error handling for elements that get to PDF code without their required children. The ones that come quickly to mind are `dl`, `ol`, `ul`, `sl`, and tables.

**Comment by raducoravu** [comment]
I agree.

**Comment by robander** [comment]
With that in mind - and the conref test case seeming unreliable - I found it easier to make a test case that just filters out required elements in the obvious locations. I'm not surprised these fail with FOP because it results in FO containers without required children. Just a little surprised that it also fails HTML5 / XHTML with:
```
 [pipeline] Failed to transform document: An empty sequence is not allowed as th
e result of template match="*[contains(@class, ' topic/simpletable ')]" mode="di
ta2html:get-max-entry-count"
```

[2701.zip](https://github.com/dita-ot/dita-ot/files/1041862/2701.zip)

**Comment by robander** [comment]
I'm working on improved fallback for XHTML, HTML5, and PDF.

Primary goal: don't crash.

Secondary goal: give reasonably not-bad output for this scenario, which really reflects bad content design. Basically, if we can tell there are no `<dlentry>` in a definition list, skip the list; if we can tell there are no `<stentry>` elements inside a simple table, skip the table.

**Comment by robander** [comment]
Fixed with #2712
